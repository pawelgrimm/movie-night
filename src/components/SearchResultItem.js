import React, { useContext, useEffect, useReducer, useState } from "react";
import ReactImageFallback from "react-image-fallback";
import { addMovie, removeMovie } from "../actions/movie";
import MovieContext from "../context/movie-context";
import MovieImage from "./MovieImage";

const SearchResultItem = ({ movie, resetSearch }) => {
  const { movies: savedMovies, dispatch } = useContext(MovieContext);
  const [button, setButton] = useState("+");
  const [buttonClass, setButtonClass] = useState("");

  useEffect(() => {
    if (savedMovies.find((savedMovie) => savedMovie.id === movie.id)) {
      toggleButtonState();
    }
  }, []);

  const toggleButtonState = () => {
    setButton(button === "+" ? "-" : "+");
    setButtonClass(buttonClass ? "" : "button--secondary");
  };

  const onClick = () => {
    toggleButtonState();
    if (button === "+") {
      dispatch(addMovie(movie));
      resetSearch();
    } else {
      dispatch(removeMovie(movie.id));
    }
  };
  return (
    <div className="search-result-item">
      <MovieImage
        title={movie.title}
        imageWidth={68}
        filePath={movie.poster_path}
        type="poster"
        className="search-result-item__poster"
      />
      <div className="search-result-item__info">
        <h3>{movie.title}</h3>
        <span>{movie.release_date}</span>
        <p>{movie.overview}</p>
      </div>
      <div className="search-result-item__button">
        <button className={"button " + buttonClass} onClick={onClick}>
          {button}
        </button>
      </div>
    </div>
  );
};

export default SearchResultItem;
