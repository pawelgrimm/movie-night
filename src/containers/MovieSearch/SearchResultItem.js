import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import MovieImage from "../../components/MovieImage/MovieImage";
import { useSearchContext } from "./search-context";
import { addAndSaveMovie, removeMovie } from "../../services/ballot/actions";
import { formatReleaseYear } from "../../services/theMovieDb/theMovieDb";

const mapStateToProps = ({ ballot }) => ({
  movies: ballot.movies,
});

const mapDispatchToProps = (dispatch) => ({
  addMovie: (movie) => {
    dispatch(addAndSaveMovie(movie.id, movie));
  },
  removeMovie: (id) => dispatch(removeMovie(id)),
});

export const SearchResultItem = ({ movie, movies, addMovie, removeMovie }) => {
  const [button, setButton] = useState("+");
  const [buttonClass, setButtonClass] = useState("");
  const { resetSearch } = useSearchContext();

  useEffect(() => {
    if (movies.find((id) => id === movie.id)) {
      toggleButtonState();
    }
  }, []);

  const toggleButtonState = () => {
    setButton(button === "+" ? "-" : "+");
    setButtonClass(buttonClass ? "" : "button--secondary");
  };

  const onClick = () => {
    toggleButtonState();
    if (button === "+") {
      addMovie(movie);
    } else {
      removeMovie(movie.id);
    }
    resetSearch();
  };
  return (
    <div className="search-result-item" tabIndex={0}>
      <MovieImage
        title={movie.title}
        imageWidth={68}
        filePath={movie.poster_path}
        type="poster"
        className="search-result-item__poster"
      />
      <div className="search-result-item__info">
        <h3>{movie.title}</h3>
        <span>{formatReleaseYear(movie.release_date)}</span>
        <p>{movie.overview}</p>
      </div>
      <div className="search-result-item__button">
        <button className={"button " + buttonClass} onClick={onClick}>
          {button}
        </button>
      </div>
    </div>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultItem);
