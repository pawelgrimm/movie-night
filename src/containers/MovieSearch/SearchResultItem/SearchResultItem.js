import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import MovieImage from "../../../components/MovieImage/MovieImage";
import { useSearchContext } from "../search-context";
import { addAndSaveMovie, removeMovie } from "../../../services/ballot/actions";
import { formatReleaseYear } from "../../../utils/theMovieDb";
import style from "./search-result-item.module.scss";
import Button from "../../../components/Button/Button";

const mapStateToProps = ({ ballot }) => ({
  movies: ballot.movies,
});

const mapDispatchToProps = (dispatch) => ({
  addMovie: (movie) => {
    dispatch(addAndSaveMovie(movie.id, movie));
  },
  removeMovie: (id) => dispatch(removeMovie(id)),
});

export const SearchResultItem = ({ movie, movies, addMovie, removeMovie }) => {
  const [button, setButton] = useState("+");
  const [isSecondary, setIsSecondary] = useState(false);
  const { resetSearch } = useSearchContext();

  useEffect(() => {
    if (movies.find((id) => id === movie.id)) {
      toggleButtonState();
    }
  }, []);

  const toggleButtonState = () => {
    setButton(button === "+" ? "-" : "+");
    setIsSecondary((prev) => !prev);
  };

  const onClick = () => {
    toggleButtonState();
    if (button === "+") {
      addMovie(movie);
    } else {
      removeMovie(movie.id);
    }
    resetSearch();
  };
  return (
    <div className={style.resultItem} tabIndex={0}>
      <MovieImage
        title={movie.title}
        imageWidth={68}
        filePath={movie.poster_path}
        type="poster"
        className={style.poster}
      />
      <div className={style.info}>
        <h3>{movie.title}</h3>
        <span>{formatReleaseYear(movie.release_date)}</span>
        <p>{movie.overview}</p>
      </div>
      <div className={style.buttons}>
        <Button type={isSecondary} className={style.button} onClick={onClick}>
          {button}
        </Button>
      </div>
    </div>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultItem);
