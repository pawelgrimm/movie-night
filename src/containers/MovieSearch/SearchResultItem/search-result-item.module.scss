@import "../../../global/styles/base/settings";

.result-item {
  display: flex;
  flex-direction: row;
  align-content: center;
  transition: background 300ms ease;
  padding: $s-size;
  border-radius: 3px;
  margin: 0 $s-size;
  &:hover,
  &:focus-within {
    background: lighten($card-background, 20%);
    outline: none;
  }
}

.poster {
  align-self: center;
  flex: 0 0 68px;
  margin: 0 $xs-size 0 0;
  border-radius: 1px;
  @media (max-width: $desktop-breakpoint) {
    flex: 0 0 45px;
  }
}

.info {
  flex-grow: 1;
  justify-items: left;
  margin-right: 0;
  * {
    line-height: 1;
    margin: 0;
    font-weight: 300;
  }
  span {
    line-height: 1;
    font-size: $font-size-small;
    color: $grey;
  }
}

.buttons {
  align-self: center;
  flex: 0 0;
  margin: 0 0 0 $xs-size;
}

.button {
  width: unset;
  margin: 0;
}
