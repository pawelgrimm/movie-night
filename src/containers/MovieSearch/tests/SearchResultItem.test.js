import { expect } from "@jest/globals";
import { shallow } from "enzyme";
import React from "react";
import { SearchResultItem } from "../SearchResultItem";
import { movies } from "../../../global/tests/fixtures/fixtures";

test("should render search result item correctly", () => {
  const movie = movies[2];
  const wrapper = shallow(
    <SearchResultItem
      movies={[]}
      addMovie={() => {}}
      removeMovie={() => {}}
      movie={movie}
      resetSearch={() => {}}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
