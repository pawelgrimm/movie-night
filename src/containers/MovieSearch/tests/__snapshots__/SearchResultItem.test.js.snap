// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render search result item correctly 1`] = `
<div
  className="resultItem"
  tabIndex={0}
>
  <MovieImage
    className="poster"
    filePath="/gycdE1ARByGQcK4fYR2mgpU6OO.jpg"
    imageWidth={68}
    title="Before Sunset"
    type="poster"
  />
  <div
    className="info"
  >
    <h3>
      Before Sunset
    </h3>
    <span>
      2004
    </span>
    <p>
      Nine years ago two strangers met by chance and spent a night in Vienna that ended before sunrise. They are about to meet for the first time since. Now they have one afternoon to find out if they belong together.
    </p>
  </div>
  <div
    className="button"
  >
    <Button
      onClick={[Function]}
      type={false}
    >
      +
    </Button>
  </div>
</div>
`;
